pkg/services/sqlstore/bulk.go

```diff
@@ -0,0 +1,59 @@
+package sqlstore
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+)
+
+const DefaultBatchSize = 1000
+
+type BulkOpSettings struct {
+	BatchSize int
+}
+
+func NativeSettingsForDialect(d migrator.Dialect) BulkOpSettings {
+	return BulkOpSettings{
+		BatchSize: d.BatchSize(),
+	}
+}
+
+func normalizeBulkSettings(s BulkOpSettings) BulkOpSettings {
+	if s.BatchSize < 1 {
+		s.BatchSize = DefaultBatchSize
+	}
+	return s
+}
+
+func (sess *DBSession) BulkInsert(table interface{}, recordsSlice interface{}, opts BulkOpSettings) (int64, error) {
```

Comment:
Accepting `interface{}` for the records is a necessary evil:
- It's consistent with xorm's API. This reflection-based approach is what xorm already does under the hood.
- Slices aren't covariant in Go, so accepting a `[]interface{}` forces the caller to convert from the type they are using, which is a risk for unneeded copies. Here, I focused on minimizing copies.
- Generics are an _okay_ solution, but can't be used on methods. The method approach offered a cleaner API, was more discoverable, and feels more native to current xorm users.